assemblyscript/assemblyscript.go

```diff
@@ -0,0 +1,182 @@
+package assemblyscript
+
+import (
+	"context"
+	"crypto/rand"
+	"encoding/binary"
+	"fmt"
+	"io"
+	"math"
+	"os"
+	"time"
+	"unicode/utf16"
+
+	"github.com/tetratelabs/wazero"
+	"github.com/tetratelabs/wazero/api"
+)
+
+// Instantiate instantiates a module implementing abort, trace, and seed for use from AssemblyScript programs.
+// The instantiated module will output abort messages to os.Stderr, not output trace messages, and use
+// crypto/rand as the source for seed values. If the AssemblyScript program is configured to use WASI, by
+// calling "import wasi" in any file, these functions will not be used.
+func Instantiate(ctx context.Context, r wazero.Runtime) (api.Closer, error) {
+	return NewModuleBuilder().Instantiate(ctx, r)
+}
+
+// ModuleBuilder allows configuring the module that will export functions used automatically by AssemblyScript.
+type ModuleBuilder interface {
+	// WithAbortWriter sets the io.Writer that should be used to write abort messages to.
+	// Abort messages are written when an error is thrown within an AssemblyScript program.
+	//
+	// Defaults to os.Stderr.
+	WithAbortWriter(writer io.Writer) ModuleBuilder
+
+	// WithTraceWriter sets the io.Writer that should be used to write trace messages to.
+	// Trace messages are written when the built-in trace function is called within an
+	// AssemblyScript program.
+	//
+	// Defaults to io.Discard meaning trace messages are not written.
+	WithTraceWriter(writer io.Writer) ModuleBuilder
```

Comment:
Similarly, this shouldn't be assignable here as it is underneath the modules importing it. It should be io.Discard by default and allowed only to set to `WithTraceToStdout() or StdErr()` anything else should be an explicit user request as it complicates sandboxing as the real owner is the importing module.